src/extra/utils/ftd.pyx

```diff
@@ -0,0 +1,517 @@
+                    cfd_j = fraction * signal[j] - _sinc_interp(
+                        signal, <data_t>j - delay)
+                else:
+                    cfd_i = _sinc_interp(
+                        signal, <data_t>i - delay) - fraction * signal[i]
+                    cfd_j = _sinc_interp(
+                        signal, <data_t>j - delay) - fraction * signal[j]
+            elif negative:
+                cfd_i = fraction * signal[i] - signal[i - int_delay]
+                cfd_j = fraction * signal[j] - signal[j - int_delay]
+            else:
+                cfd_i = signal[i - int_delay] - fraction * signal[i]
+                cfd_j = signal[j - int_delay] - fraction * signal[j]
+
+            # From this point on, the CFD values are computed in such a
+            # way as if the signal and threshold would be positive.
+
+            if cfd_i < walk and cfd_j > walk and i > next_edge:
+                if interp == EdgeInterpolation.NEAREST:
+                    edge_pos = i if fabs(cfd_j - walk) > fabs(cfd_i - walk) \
+                        else j
+                elif interp == EdgeInterpolation.LINEAR:
+                    edge_pos = i + (cfd_i - walk) / (cfd_i - cfd_j)
+                elif interp == EdgeInterpolation.SPLINE:
+                    raise NotImplementedError('spline interpolation')
+                elif interp == EdgeInterpolation.SINC:
+                    # As sinc interpolation still makes use of the
+                    # original (possibly negative!) signal, it also
+                    # requires the original walk value for comparions.
+                    edge_pos = _cfd_sinc_interp(signal, i, j, delay, fraction,
+                                                orig_walk)
+                else:
+                    raise ValueError('invalid interpolation mode')
+
+                edges[edge_idx] = edge_pos
+                next_edge = i + width
+                edge_idx += 1
+
+                if edge_idx == max_edge:
+                    break
+
+    return np.asarray(edges)[:edge_idx], np.asarray(amplitudes)[:edge_idx], \
+        edge_idx
+
+
+def dled(
+    data_array_t signal, data_t threshold, data_t ratio_max=0.6, data_t width=0,
+    int interp = EdgeInterpolation.LINEAR,
+    data_array_t edges = None, data_array_t amplitudes = None
+):
+    """Dynamic leading edge discriminator.
+
+    This algorithm finds the leading edge of pulses in a 1D signal
+    located at a certain ratio to the pulse's peak value.
+
+    As with the constant fraction discriminator, it is assumed the
+    rising pulse slope points away from zero.
+
+    Args:
+        signal (ArrayLike): 1D input array with analog signal.
+        threshold (data-type): Trigger threshold, positive values imply
+            a positive pulse slope while negative values correspondingly
+            imply a negative pulse slope.
+        ratio_max (data-type, optional): Ratio of leading edge to peak
+            value, 0.6 by default.
```

Comment:
This should presumably be between 0 & 1? Is it worth saying that in the docstring & maybe even checking it in the function?

Reply:
Good idea to clarify, and a C-level check will be as good as free. Funny things would happen if you go negative, and nothing would happen if > 1.

Reply:
Done.